lib/puppet/util/package/version/pip.rb

```diff
@@ -0,0 +1,230 @@
+      @post_data    = parse_letter_version(matched[:post_l], matched[:post_n1] || matched[:post_n2]) if matched[:post_l] || matched[:post_n1] || matched[:post_n2]
+      @dev_data     = parse_letter_version(matched[:dev_l], matched[:dev_n])                         if matched[:dev_l]  || matched[:dev_n]
+      @local_data   = parse_local_version(matched[:local])                                           if matched[:local]
+
+      @key = compose_key(@epoch_data, @release_data, @pre_data, @post_data, @dev_data, @local_data)
+    end
+
+    def self.parse(version)
+      new(version)
+    end
+
+    def self.compare(version_a, version_b)
+      version_a = new(version_a) unless version_a.is_a?(self.class)
+      version_b = new(version_b) unless version_a.is_a?(self.class)
+
+      version_a <=> version_b
+    end
+
+    def to_s
+      parts = []
+
+      parts.push("#{epoch}!")                   if epoch
+      parts.push(release.join("."))             if release
+      parts.push(pre_release)                   if pre_release
+      parts.push(".post#{post_release[1]}")     if post_release
+      parts.push(".dev#{dev_release[1]}")       if dev_release
+      parts.push("+#{local_version.join(".")}") if local_version
+
+      parts.join
+    end
+
+    def <=>(other)
+      raise ValidationFailure.new(other.to_s) unless other.is_a?(self.class)
+      compare(self.key, other.key)
+    end
+
+    def base_version
+      to_s.split('+').first
+    end
+
+    def major_version
+      release[0].to_i
+    end
+
+    def minor_version
+      release[1].to_i
+    end
+
+    def micro_version
+      release[2].to_i
+    end
+
+    def is_pre_release?
+      dev_release != nil || pre_release != nil
+    end
+
+    def is_post_release?
+      post_release != nil
+    end
+
+    def is_dev_release?
+      dev_release != nil
+    end
+
+    def key
+      @key
+    end
+
+    private
+
+    @key = nil
+
+    def key=(val)
+      @key = val
+    end
+
+    def epoch
+      @epoch_data if @epoch_data != 0 && @epoch_data
+    end
+
+    def release
+      @release_data
+    end
+
+    def pre_release
+      @pre_data.join if @pre_data
+    end
+
+    def post_release
+      @post_data
+    end
+
+    def dev_release
+      @dev_data
+    end
+
+    def local_version
+      @local_data
+    end
+
+    def parse_letter_version(letter, number)
+      if letter
+        number = 0 if !number
+        letter.downcase!
+
+        if letter == "alpha"
+          letter = "a"
+        else
+          if letter == "beta"
+            letter = "b"
+          else
+            if ["c", "pre", "preview"].include?(letter)
+              letter = "rc"
+            else
+              if ["rev", "r"].include?(letter)
+                letter = "post"
+              end
+            end
+          end
+        end
+
+        return [letter, number.to_i]
+      end
+
+      if !letter && number
+        letter = "post"
+        return [letter, number.to_i]
+      end
```

Comment:
if there is no else, is better suited to use a guard clause:
https://anthonygharvey.com/ruby/guard_clauses_vs_nested_if_statements